web/@types/styled-components.d.ts

```diff
@@ -16,9 +16,18 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import 'styled-components';
+import { CSSProp } from 'styled-components';
+import 'react';
+
 import { Theme } from 'design/theme/themes/types';
 
 declare module 'styled-components' {
   export interface DefaultTheme extends Theme {}
 }
+
+// https://styled-components.com/docs/api#usage-with-typescript
+declare module 'react' {
+  interface Attributes {
+    css?: CSSProp | undefined;
+  }
+}
```

Comment:
The docs point to the newest version of styled-components. Is this usage correct in the context of the version that we use?

Reply:
Yeah, the docs for CSS prop for our version say this:
<img width="812" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/f77195a4-2166-448a-b338-f8f9227a134d">

I prefer the module augmentation option as it is more explicit.

Regarding the theme: I tried to pass our theme type in a way they show it, but it didn't make statements like that typed:
```ts
declare module 'react' {
  interface Attributes {
    css?: CSSProp<Theme> | undefined;
  }


css={`
   color: ${props => props.theme...}; // theme is still any
`}
```
I'm not sure what it doesn't work, but since the type defaults to `DefaultTheme`
 (`export type CSSProp<T = AnyIfEmpty<DefaultTheme>>`) in the end I didn't pass anything to this type.
Maybe it's broken on this version, idk.

Reply:
Hmm, indeed, something's not right there. It works when you explicitly call `css` with the interpolated string. Then it interprets `props.theme` as `DefaultTheme` and the rest works correctly because Bartosz extended that interface some time ago.

Someone complained about this problem in a long-standing issue about `CSSProp` in general: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31245#issuecomment-474232597.

I think the culprit is here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3c78f0fd3c1bd7c16bde2b58e3dccadfccbdb0f1/types/styled-components/index.d.ts#L427-L428

If you remove `string` from that type, and then look here:

https://github.com/gravitational/teleport/blob/350dbe3e3109daae08409c0a164c2405f1b7be9e/web/packages/teleterm/src/ui/Vnet/VnetConnectionItem.tsx#L265-L268

You'll see that it complains that `string` is not assignable to `CSSProp`. In other words, it interprets that string as a string rather than `FlattenInterpolation<ThemeProps<T>>`. The `css` function doesn't seem to have that problem, (here's [its type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3c78f0fd3c1bd7c16bde2b58e3dccadfccbdb0f1/types/styled-components/index.d.ts#L262-L274)).

Reply:
Thanks for the investigation, I couldn't find other users' reports.